File: leetcode/medium/1630. Arithmetic Subarrays/solution.js
// https://leetcode.com/problems/arithmetic-subarrays/
/**
 * @param {number[]} nums
 * @param {number[]} l
 * @param {number[]} r
 * @return {boolean[]}
 */
function checkArithmeticSubarrays(nums, l, r) {
    const canRearrange = (subArr) => {
        subArr.sort((a, b) => a - b);
        const diff = subArr[1] - subArr[0];
        for (let i = 0; i < subArr.length - 1; i += 1) {
            if (subArr[i + 1] - subArr[i] !== diff) return false;
        }
        return true;
    }

    return l.map((_, i) => canRearrange(nums.slice(l[i], r[i] + 1)));
}

// checkArithmeticSubarrays([4, 6, 5, 9, 3, 7], [0, 0, 2], [2, 3, 5])
console.log(checkArithmeticSubarrays([-12,-9,-3,-12,-6,15,20,-25,-20,-15,-10], [0,1,6,4,8,7], [4,4,9,7,9,10]))

File: leetcode/medium/1630. Arithmetic Subarrays/solution.cc
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    vector<bool> checkArithmeticSubarrays(vector<int>& nums, vector<int>& l, vector<int>& r) {
        vector<bool> res;
        for (int i = 0; i < l.size(); i += 1) {
            vector<int> sliced(nums.begin() + l[i], nums.begin() + r[i] + 1);
            res.push_back(canRearrange(sliced));
        }
        return res;
    }

    bool canRearrange(vector<int>& subArr) {
        sort(subArr.begin(), subArr.end());
        int diff = subArr[1] - subArr[0];

        for (int i = 0; i < subArr.size() - 1; i += 1) {
            if (subArr[i+1] - subArr[i] != diff) return false;
        }
        return true;
    }
};

int main() {
    Solution solution;
    vector<int> nums = {4, 6, 5, 9, 3, 7}, l = {0, 0, 2}, r = {2, 3, 5};
    vector<bool> res = solution.checkArithmeticSubarrays(nums, l, r);

    for (int i = 0; i < res.size(); i += 1) {
        cout << res[i] << endl;
    }
    return 0;
}